### DISCUSSION TITLE
Plugin which injects POST routes

### DISCUSSION THREAD

**Discussion by svoop** [jeremyevans/roda#320] [General]
Hiya

I'm playing around with Roda plugins, however, I'm not sure what's the recommended way to implement the following simple case: a plugin which injects POST routes.

Given the following plugin:

```ruby
class Roda
  module RodaPlugins
    module GithubWebhook
      module RequestMethods

        def github_webhook
          'been there'
        end

      end
    end

    register_plugin :github_webhook, GithubWebhook
  end
end
```

Used in an app like so:

```ruby
plugin :github_webhook

route do |r|
  r.github_webhook
end
```

With this in place `GET /github_webhook` responds with "been there".

How can I implement the same for a route such as `POST /github/webhook`?

(Since the webhook is automatically configured upstream, the route `/github/webhook` is mandatory. I'm looking for a way to add the `POST /github/webhook` route to the Roda app with minimal code added to the application.rb while still doing things the Roda way.)

Thanks a lot for your help!

**Comment by janko** [comment]
The `r` local variable is the request object that is also available via the `request` method. So you can do:

```rb
def github_webhook
  request.post "github/webhook" do
    'been there'
  end
end
```

**Reply by svoop** [reply]
That'd be really awesome, but it's not available – for starters testing with `GET`:

```ruby
class Roda
  module RodaPlugins
    module GithubWebhook
      module RequestMethods

        def github_webhook
          request.get "github/webhook" do
            'been there'
          end
        end

      end
    end

    register_plugin :github_webhook, GithubWebhook
  end
end
```

This raises the following error on line 7 above:

> undefined local variable or method `request' for #<Application::RodaRequest GET /github/webhook> (NameError)

Apparently, there's still a bit missing.

**Reply by jeremyevans** [reply]
Since you are in `RequestMethods`, the receiver is the request.  So you can do:

```ruby
        def github_webhook
          get "github/webhook" do
            'been there'
          end
        end
```

**Reply by svoop** [reply]
Awesome, really neat and concise, thanks a bunch @jeremyevans and @janko !
